### DISCUSSION TITLE
text/template.Execute: non-concrete value string

### DISCUSSION THREAD

**Discussion by networkop** [cue-lang/cue#1832] [Q&A] [answered]
I'm trying to read Go template file from a disk and instantiate it with cue values but I get the above error. Am I doing something wrong?

Here's my `go.tmpl` file:
```tmpl
VALUE: {{ .x }}
```
And my `cue_tool.cue` file:
```cue
package main

import (
	"tool/cli"
	"text/template"
	"tool/file"
)

file_template: file.Read & {
	filename: "go.tmpl"
	contents: string
}

v: x: "FOO"

foo: """
	VALUE: {{ .x }}
	"""

cli_commands: template.Execute(file_template.contents, v)

command: test: {
	print: cli.Print & {
		text: cli_commands
	}
}
```

When I run `cue test` I get the following:

```
$ cue test
command.test.print.text: invalid string argument: error in call to text/template.Execute: non-concrete value string:
    ./cue_tool.cue:23:2
    ./cue_tool.cue:11:12
    ./cue_tool.cue:20:15
    ./cue_tool.cue:24:3
    tool/cli:4:3
```

If I replace the `file_template.contents` with `foo` in the above cue file, I get the result I expected:
```
$ cue test
VALUE: FOO
```

It seems similar to https://github.com/cue-lang/cue/discussions/1122 and #1095 but I'm running cue version `0.4.3` which should include the fix.

**Comment by slewiskelly** [comment, marked as answer]
The following works: 

```go
// go.tmpl
VALUE: {{ .x }}
```

```cue
// test_tool.cue
package main

import (
	"tool/cli"
	"text/template"
	"tool/file"
)

v: x: "FOO"

command: test: {
	file_template: file.Read & {
		filename: "go.tmpl"
		contents: string
	}

	print: cli.Print & {
		text: template.Execute(file_template.contents, v)
	}
}
```

The `file.Read` task needs to be in the context of the command otherwise the file isn't read and `contents` is an incomplete value.

**Reply by networkop** [reply]
brilliant, thanks @slewiskelly, it worked!
it still feels a bit strange that it only affects `text/template` but if I used `file_template.contents` directly in `cli.Print.text` it works even when I read the file outside of the context of the command.

**Reply by myitcv** [reply]
@slewiskelly definitely has the answer which will always work today. Represented here as a [self-contained reproducer](https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers):

```
exec cue cmd test
cmp stdout stdout.golden

-- test_tool.cue --
package main

import (
	"tool/cli"
	"text/template"
	"tool/file"
)

v: x: "FOO"

command: test: {
	file_template: file.Read & {
		filename: "go.tmpl"
		contents: string
	}

	print: cli.Print & {
		text: template.Execute(file_template.contents, v)
	}
}
-- go.tmpl --
VALUE: {{ .x }}
-- stdout.golden --
VALUE: FOO

```




> it still feels a bit strange that it only affects `text/template` but if I used `file_template.contents` directly in `cli.Print.text` it works even when I read the file outside of the context of the command.

This is a good observation, @networkop, and points to one of the features we will implement as part of reworking `cue cmd` in https://github.com/cue-lang/cue/issues/1325. Namely, that implicit task dependencies should always be treated as tasks themselves to be run. I suspect the reason that's not happening via `cli_commands` is that the transitive dependency on `file_template` (a task) is lost/dropped through the call to `template.Execute`. So I'll capture that in a bug report, because I don't think we have an issue for this particular "lost" dependency.

**Reply by myitcv** [reply]
Raised as https://github.com/cue-lang/cue/issues/1834.

Thanks again @networkop and @slewiskelly!
